x/twap/logic_test.go

```diff
@@ -42,76 +43,287 @@ func newExpRecord(accum0, accum1 sdk.Dec) types.TwapRecord {
+func (s *TestSuite) TestNewTwapRecord() {
+	// prepare pool before test
+	poolId := s.PrepareBalancerPoolWithCoins(defaultTwoAssetCoins...)
+
+	tests := map[string]struct {
+		poolId        uint64
+		denom0        string
+		denom1        string
+		expectedErr   bool
+		expectedPanic bool
+	}{
+		"denom with lexicographical order": {
+			poolId,
+			denom1,
+			denom0,
+			false,
+			false,
+		},
+		"denom with non-lexicographical order": {
+			poolId,
+			denom0,
+			denom1,
+			false,
+			false,
+		},
+		"new record with same denom": {
+			poolId,
+			denom0,
+			denom0,
+			true,
+			false,
+		},
+		"non existent pool id": {
+			poolId + 1,
+			denom1,
+			denom0,
+			false,
+			true,
+		},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			if test.expectedPanic {
+				s.Require().Panics(func() {
+					twap.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, test.poolId, test.denom0, test.denom1)
+				})
+				return
+			}
+
+			twapRecord, err := twap.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, test.poolId, test.denom0, test.denom1)
+
+			if test.expectedErr {
+				s.Require().Error(err)
+			} else {
+				s.Require().NoError(err)
+
+				s.Require().Equal(test.poolId, twapRecord.PoolId)
+				s.Require().Equal(s.Ctx.BlockHeight(), twapRecord.Height)
+				s.Require().Equal(s.Ctx.BlockTime(), twapRecord.Time)
+				s.Require().Equal(sdk.ZeroDec(), twapRecord.P0ArithmeticTwapAccumulator)
+				s.Require().Equal(sdk.ZeroDec(), twapRecord.P1ArithmeticTwapAccumulator)
+			}
+
+		})
+	}
+}
+
+func (s *TestSuite) TestUpdateRecord() {
 	tests := map[string]struct {
-		record          types.TwapRecord
-		interpolateTime time.Time
-		expRecord       types.TwapRecord
+		inputRecord       types.TwapRecord
+		updateBlockHeight bool
+		updateSpotPrice   bool
+		updateBlockTime   bool
+		expRecord         types.TwapRecord
 	}{
-		"0accum": {
-			record:          newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
-			interpolateTime: time.Unix(2, 0),
+		"happy path, zero accumulator": {
+			inputRecord:       newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateBlockHeight: false,
+			updateSpotPrice:   false,
+			expRecord:         newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+		"update block height": {
+			inputRecord:       newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateBlockHeight: true,
+			expRecord:         newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+		"update block time": {
+			inputRecord:     newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateSpotPrice: true,
+			updateBlockTime: true,
 			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
 		},
+		"spot price changed": {
+			inputRecord:     newRecord(1, s.Ctx.BlockTime(), sdk.NewDec(10), zeroDec, zeroDec),
+			updateSpotPrice: true,
+			expRecord:       newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
```

Comment:
it seems like we aren't using the expRecord in the below tests at all

Reply:
yeah the original intention was to use it at some point, but changed to testing it using other states for check! Deleted this field